### ISSUE TITLE
ansible Sharing connection will report error

### ISSUE THREAD

**Issue by targetoyes** [ansible/ansible#12919]
##### Issue Type:

  - Bug Report

##### Ansible Version:

tried several versions(1.9.*)
not sure ，maybe all version

##### Ansible Configuration:

default settings

##### Environment:

Ubuntu 12.04, centos5/6(just test these)

##### Summary:

SSH Error: Shared connection to test closed

##### Steps To Reproduce:

like issue ,for example: `ansible test -a "echo 1"& ansible test -a "echo 1"`

##### Expected Results:
 not always act like this
```
10.8.17.27 | success | rc=0 >>
1

10.8.17.27 | success | rc=0 >>
1
```
##### Actual Results:
```
test | FAILED => SSH Error: Shared connection to test closed.
It is sometimes useful to re-run the command using -vvvv, which prints SSH debug output to help diagnose the issue.
test | FAILED => SSH Error: Shared connection to test closed.
It is sometimes useful to re-run the command using -vvvv, which prints SSH debug output to help diagnose the issue.
```

I know this situation may not always happen,but i wrote it in my program, however I don't like this to happen at any time.
I just set a Interval to sleep to avoid this problem now.Maybe we can check other tasks whether is competing resource before excuting this task, and then we decide whether to wait or not.

**Comment by bcoca** [comment]
Can You Help Us Out?
====================

Thanks for filing a ticket! I am the friendly GitHub Ansibot.

It looks like you might not have filled out the issue description based on our standard issue template.  You might not have known about that, and that's ok too, we'll tell you how to do it.

We have a standard template because Ansible is a really busy project and it helps to have some standard information in each ticket, and GitHub doesn't yet provide a standard facility to do this like some other bug trackers.  We hope you understand as this is really valuable to us!.

Solving this is simple: please copy the contents of this [template](https://raw.githubusercontent.com/ansible/ansible/devel/ISSUE_TEMPLATE.md) and **paste it into the description** of your ticket.  That's it!

If You Had A Question To Ask Instead
====================================

If you happened to have a "how do I do this in Ansible" type of question, that's probably more of a user-list question than a bug report, and you should probably ask this question on the project mailing list instead.

However, if you think you have a bug, the report is the way to go!  We definitely want all the bugs filed :)  Just trying to help!

About Priority Tags
===================

Since you're here, we'll also share some useful information at this time.

In general tickets will be assigned a priority between P1 (highest) and P5, and then worked in priority order. We may also have some follow up questions along the way, so keeping up with follow up comments via GitHub notifications is a good idea.

Due to large interest in Ansible, humans may not comment on your ticket immediately.

Mailing Lists
=============

If you have concerns or questions, you're welcome to stop by the ansible-project or ansible-development mailing lists, as appropriate.  Here are the links:

  * https://groups.google.com/forum/#!forum/ansible-project - for discussion of bugs and how-to type questions
  * https://groups.google.com/forum/#!forum/ansible-devel - for discussion on how to implement a code change, or feature brainstorming among developers

Thanks again for the interest in Ansible!

**Comment by targetoyes** [comment]
##### Issue Type:

  - Bug Report

##### Ansible Version:

tried several versions(1.9.*)
not sure ，maybe all version

##### Ansible Configuration:

default settings

##### Environment:

Ubuntu 12.04, centos5/6(just test these)

##### Summary:

SSH Error: Shared connection to test closed

##### Steps To Reproduce:

like issue ,for example: ansible test -a "echo 1"& ansible test -a "echo 1"

##### Expected Results:
 not always act like this

10.8.17.27 | success | rc=0 >>
1

10.8.17.27 | success | rc=0 >>
1

##### Actual Results:

test | FAILED => SSH Error: Shared connection to test closed.
It is sometimes useful to re-run the command using -vvvv, which prints SSH debug output to help diagnose the issue.
test | FAILED => SSH Error: Shared connection to test closed.
It is sometimes useful to re-run the command using -vvvv, which prints SSH debug output to help diagnose the issue.




I know this situation may not always happen,but i wrote it in my program, however I don't like this to happen at any time.

**Comment by bcoca** [comment]
seems to work in 2.0

```
 ansible nyx -a "echo 1"& ansible nyx -a "echo 1"                                                                                                                                
[1] 28199                                                                                                                                                                                              
nyx | SUCCESS | rc=0 >>                                                                                                                                                                                
1

nyx | SUCCESS | rc=0 >>
1

[1]+  Done                    ansible nyx -a "echo 1"
```

**Comment by targetoyes** [comment]
How to download version 2.0? I can't find it in pypi or other release.

**Comment by targetoyes** [comment]
I had tried version 2.0.0.1,  but failed like the 1.9.x.x, and had three results.
Because of using module of copy to send files to the same remote server at the same time,this problem may cause interruption.

```
deploy@ansible:/etc/apt$ ansible --version
ansible 2.0.0.1
  config file = /etc/ansible/ansible.cfg
  configured module search path = /usr/share/ansible
deploy@ansible:/etc/apt$ ansible 10.8.17.27 -a 'echo 1' &ansible 10.8.17.27 -a 'echo 1'
[1] 3053
10.8.17.27 | FAILED | rc=0 >>
MODULE FAILURE

10.8.17.27 | FAILED | rc=0 >>
MODULE FAILURE

[1]+  退出 2                ansible 10.8.17.27 -a 'echo 1'




deploy@ansible:/etc/apt$ ansible 10.8.17.27 -a 'echo 1' &ansible 10.8.17.27 -a 'echo 1'
[2] 3013
10.8.17.27 | UNREACHABLE! => {
    "changed": false, 
    "msg": "ERROR! SSH encountered an unknown error during the connection. We recommend you re-run the command using -vvvv, which will enable SSH debugging output to help diagnose the issue", 
    "unreachable": true
}
10.8.17.27 | UNREACHABLE! => {
    "changed": false, 
    "msg": "ERROR! SSH encountered an unknown error during the connection. We recommend you re-run the command using -vvvv, which will enable SSH debugging output to help diagnose the issue", 
    "unreachable": true
}
[1]   完成                  ansible 10.8.17.27 -a 'echo 1'




deploy@ansible:/etc/apt$ ansible 10.8.17.27 -a 'echo 1' &ansible 10.8.17.27 -a 'echo 1'
[1] 2981
10.8.17.27 | SUCCESS | rc=0 >>
1

10.8.17.27 | SUCCESS | rc=0 >>
1

```

**Comment by bcoca** [comment]
tried it several times, i don't get an error, can it be you just have connection or resource issues?

```
bcoca@erebos:~/work/ansible/docsite/rst(devel) $ ansible nyx -a 'echo 1'
nyx | SUCCESS | rc=0 >>
1

bcoca@erebos:~/work/ansible/docsite/rst(devel) $ ansible nyx -a 'echo 1'& ansible nyx -a 'echo 1'
[1] 22678
nyx | SUCCESS | rc=0 >>
1

nyx | SUCCESS | rc=0 >>
1

bcoca@erebos:~/work/ansible/docsite/rst(devel) $ ansible nyx -a 'echo 1'& ansible nyx -a 'echo 1'
[2] 22741
nyx | SUCCESS | rc=0 >>
1

[1]   Done                    ansible nyx -a 'echo 1'
bcoca@erebos:~/work/ansible/docsite/rst(devel) $ nyx | SUCCESS | rc=0 >>
1

ansible nyx -a 'echo 1'& ansible nyx -a 'echo 1'
[3] 22775
nyx | SUCCESS | rc=0 >>
1

nyx | SUCCESS | rc=0 >>
1

[2]   Done                    ansible nyx -a 'echo 1'
[3]+  Done                    ansible nyx -a 'echo 1'
bcoca@erebos:~/work/ansible/docsite/rst(devel) $ ansible nyx -a 'echo 1'& ansible nyx -a 'echo 1'
[1] 22806
nyx | SUCCESS | rc=0 >>
1

nyx | SUCCESS | rc=0 >>
1

[1]+  Done                    ansible nyx -a 'echo 1'
bcoca@erebos:~/work/ansible/docsite/rst(devel) $ 
bcoca@erebos:~/work/ansible/docsite/rst(devel) $ ansible nyx -a 'echo 1'& ansible nyx -a 'echo 1'
[1] 22840
nyx | SUCCESS | rc=0 >>
1

nyx | SUCCESS | rc=0 >>
1

[1]+  Done                    ansible nyx -a 'echo 1'
```

**Comment by targetoyes** [comment]
```
Loaded callback minimal of type stdout, v2.0
Loaded callback minimal of type stdout, v2.0
<10.8.17.27> ESTABLISH SSH CONNECTION FOR USER: None
<10.8.17.27> ESTABLISH SSH CONNECTION FOR USER: None
<10.8.17.27> SSH: EXEC ssh -C -vvv -o ControlMaster=auto -o ControlPersist=60s -o Port=22 -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o ConnectTimeout=10 -o ControlPath=/home/deploy/.ansible/cp/ansible-ssh-%h-%p-%r -tt 10.8.17.27 '( umask 22 && mkdir -p "$( echo $HOME/.ansible/tmp/ansible-tmp-1452749402.25-146280161331306 )" && echo "$( echo $HOME/.ansible/tmp/ansible-tmp-1452749402.25-146280161331306 )" )'
<10.8.17.27> SSH: EXEC ssh -C -vvv -o ControlMaster=auto -o ControlPersist=60s -o Port=22 -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o ConnectTimeout=10 -o ControlPath=/home/deploy/.ansible/cp/ansible-ssh-%h-%p-%r -tt 10.8.17.27 '( umask 22 && mkdir -p "$( echo $HOME/.ansible/tmp/ansible-tmp-1452749402.25-102946557303912 )" && echo "$( echo $HOME/.ansible/tmp/ansible-tmp-1452749402.25-102946557303912 )" )'
10.8.17.27 | UNREACHABLE! => {
    "changed": false, 
    "msg": "ERROR! SSH encountered an unknown error. The output was:\nOpenSSH_5.9p1 Debian-5ubuntu1.7, OpenSSL 1.0.1 14 Mar 2012\ndebug1: Reading configuration data /etc/ssh/ssh_config\r\ndebug1: /etc/ssh/ssh_config line 19: Applying options for *\r\ndebug1: auto-mux: Trying existing master\r\ndebug2: fd 3 setting O_NONBLOCK\r\ndebug2: mux_client_hello_exchange: master version 4\r\ndebug3: mux_client_request_forwards: requesting forwardings: 0 local, 0 remote\r\ndebug3: mux_client_request_session: entering\r\ndebug3: mux_client_request_alive: entering\r\ndebug3: mux_client_request_alive: done pid = 22082\r\ndebug3: mux_client_request_session: session request sent\r\ndebug1: mux_client_request_session: master session id: 4\r\ndebug3: mux_client_read_packet: read header failed: Broken pipe\r\ndebug2: Control master terminated unexpectedly\r\nShared connection to 10.8.17.27 closed.\r\n", 
    "unreachable": true
}
10.8.17.27 | UNREACHABLE! => {
    "changed": false, 
    "msg": "ERROR! SSH encountered an unknown error. The output was:\nOpenSSH_5.9p1 Debian-5ubuntu1.7, OpenSSL 1.0.1 14 Mar 2012\ndebug1: Reading configuration data /etc/ssh/ssh_config\r\ndebug1: /etc/ssh/ssh_config line 19: Applying options for *\r\ndebug1: auto-mux: Trying existing master\r\ndebug2: fd 3 setting O_NONBLOCK\r\ndebug2: mux_client_hello_exchange: master version 4\r\ndebug3: mux_client_request_forwards: requesting forwardings: 0 local, 0 remote\r\ndebug3: mux_client_request_session: entering\r\ndebug3: mux_client_request_alive: entering\r\ndebug3: mux_client_request_alive: done pid = 22082\r\ndebug3: mux_client_request_session: session request sent\r\ndebug1: mux_client_request_session: master session id: 3\r\ndebug3: mux_client_read_packet: read header failed: Broken pipe\r\ndebug2: Control master terminated unexpectedly\r\nShared connection to 10.8.17.27 closed.\r\n", 
    "unreachable": true
}
```

Does the environment of system cause this?
OpenSSH_5.9p1 Debian-5ubuntu1.7, OpenSSL 1.0.1 14

**Comment by bcoca** [comment]
it looks like a connectivity issue, to factor out control persist and local locks on the socket try with `-c paramiko`

**Comment by targetoyes** [comment]
thanks，by using -c paramiko,that works!

**Comment by bcoca** [comment]
can you detail the OS of the ansible machine, the target and the versions of ssh you are running?

**Comment by targetoyes** [comment]
I tried in centos yestoday , it's ok.But i use ubuntu,it can be passed.
OS detail:Ubuntu 12.04.2 LTS \n \l  
core:Linux ansible 3.5.0-23-generic #35~precise1-Ubuntu SMP Fri Jan 25 17:13:26 UTC 2013 x86_64 x86_64 x86_64 GNU/Linux
ssh: OpenSSH_5.9p1 Debian-5ubuntu1.7, OpenSSL 1.0.1 14 Mar 2012

target is the same clone machine.

**Comment by boneskull** [comment]
I'm experiencing something similar.  While testing a playbook, connections just started timing out at initial `setup` task.  I got some debug output and tried repeating the `ssh` command.

Here's its output:

```shell
# -vvv is mine; rest was pasted from verbose ansible output
$ ssh -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=pi -o ConnectTimeout=10 -o ControlPath=/Users/boneskull/.ansible/cp/ansible-ssh-%h-%p-%r marvin.local '/bin/sh -c '"'"'( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1484445157.73-176778407015933 `" && echo ansible-tmp-1484445157.73-176778407015933="` echo $HOME/.ansible/tmp/ansible-tmp-1484445157.73-176778407015933 `" ) && sleep 0'"'"''
OpenSSH_7.3p1, LibreSSL 2.4.1
debug1: Reading configuration data /Users/boneskull/.ssh/config
debug1: /Users/boneskull/.ssh/config line 35: Applying options for *
debug1: Reading configuration data /etc/ssh/ssh_config
debug1: /etc/ssh/ssh_config line 20: Applying options for *
debug1: /etc/ssh/ssh_config line 56: Applying options for *
debug1: auto-mux: Trying existing master
debug2: fd 3 setting O_NONBLOCK
debug2: mux_client_hello_exchange: master version 4
debug3: mux_client_forwards: request forwardings: 0 local, 0 remote
debug3: mux_client_request_session: entering
debug3: mux_client_request_alive: entering
debug3: mux_client_request_alive: done pid = 80550
debug3: mux_client_request_session: session request sent
debug1: mux_client_request_session: master session id: 11
ansible-tmp-1484445157.73-176778407015933=/home/pi/.ansible/tmp/ansible-tmp-1484445157.73-176778407015933
debug3: mux_client_read_packet: read header failed: Broken pipe
debug2: Received exit status from master 0
```

The applied options from my `ssh_config`:

```
Host                    *
  ForwardAgent          yes
```

The workaround--using `-c paramiko`--seems to avoid the problem.

* * *
Ansible is v2.2.0.0 with default config.

Host is macOS Sierra; target is Raspbian.  

On the target, `sshd --version` reports `OpenSSH_6.7p1 Raspbian-5+deb8u3, OpenSSL 1.0.1t  3 May 2016`.

**Comment by Nirkus** [comment]
Hey,

when running a cluster deployment playbook against some test VM, I am also affected by this bug.
The playbook adds six cluster nodes using add_host with ansible_host set to the same test VM.
When executing several tasks on all six hosts in parallel, the control connection gets reset.

ansible error message with -vvvv:
```
fatal: [c3.cluster1.invalid]: UNREACHABLE! => {
    "changed": false, 
    "msg": "Failed to connect to the host via ssh: OpenSSH_6.7p1 Debian-5+deb8u3, OpenSSL 1.0.1t  3 May 2016\r\ndebug1: Reading configuration data /home/someuser/.ssh/config\r\ndebug1: /home/someuser/.ssh/config line 16: Applying options for *.example.invalid\r\ndebug1: Readi
ng configuration data /etc/ssh/ssh_config\r\ndebug1: /etc/ssh/ssh_config line 19: Applying options for *\r\ndebug1: auto-mux: Trying existing master\r\ndebug2: fd 3 setting O_NONBLOCK\r\ndebug2: mux_client_hello_exchange: master version 4\r\ndebug3: mux_client
_forwards: request forwardings: 0 local, 0 remote\r\ndebug3: mux_client_request_session: entering\r\ndebug3: mux_client_request_alive: entering\r\ndebug3: mux_client_request_alive: done pid = 28562\r\ndebug3: mux_client_request_session: session request sent\r\
ndebug1: mux_client_request_session: master session id: 4\r\ndebug3: mux_client_read_packet: read header failed: Broken pipe\r\ndebug2: Control master terminated unexpectedly\r\nShared connection to testvm01.example.invalid closed.\r\n", 
    "unreachable": true
}
```

ansible control machine is Debian jessie running ansible v2.2.1.0:
```
$ ansible-playbook --version
ansible-playbook 2.2.1.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```

The target machine (test VM) is running Ubuntu 14.04.5 with openssh-server 6.6p1-2ubuntu2.8.
sshd logs the followign error messages on the test VM when the control connection is reset (from auth.log):
```
Mar 22 11:19:45 testvm01 sshd[21226]: Starting session: command for someuser from 192.0.2.42 port 47690
Mar 22 11:19:45 testvm01 sshd[21226]: Starting session: command on pts/6 for someuser from 192.0.2.42 port 47690
Mar 22 11:19:45 testvm01 sshd[21226]: Starting session: command for someuser from 192.0.2.42 port 47690
Mar 22 11:19:45 testvm01 sshd[21226]: Starting session: command on pts/7 for someuser from 192.0.2.42 port 47690
Mar 22 11:19:45 testvm01 sshd[23179]: fatal: mm_request_receive_expect: read: rtype 29 != type 157
Mar 22 11:19:45 testvm01 sshd[21226]: error: buffer_get_ret: trying to get more bytes 1 than in buffer 0
Mar 22 11:19:45 testvm01 sshd[21226]: error: buffer_get_char_ret: buffer_get_ret failed
Mar 22 11:19:45 testvm01 sshd[21226]: fatal: buffer_get_char: buffer error
Mar 22 11:19:45 testvm01 sshd[21161]: fatal: mm_request_receive: read: Connection reset by peer
Mar 22 11:19:45 testvm01 sshd[21161]: pam_unix(sshd:session): session closed for user someuser
```
FYI: Running with `-f 1`or `serial: 1` does not trigger this bug.

**Comment by Nirkus** [comment]
Probably related:
[Ubuntu sshd issue #1334916](https://bugs.launchpad.net/ubuntu/+source/openssh/+bug/1334916)

**Comment by ansibot** [comment]
@targetoyes Greetings! Thanks for taking the time to open this issue. In order for the community to handle your issue effectively, we need a bit more information. 

Here are the items we could not find in your description:
- component name

Please set the description of this issue with this template:
https://raw.githubusercontent.com/ansible/ansible/devel/.github/ISSUE_TEMPLATE.md

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ghost** [comment]
Broken from 2.2 to 2.4. I've tried `-c paramiko` and `-f 1`.

**Comment by mickgortenmulder** [comment]
same problem here:

```fatal: [hostname.domain.nl]: FAILED! => {"changed": false, "failed": true, "invocation": {"module_name": "command"}, "module_stderr": "OpenSSH_7.4p1, OpenSSL 1.0.2k-fips  26 Jan 2017\r\ndebug1: Reading configuration data /etc/ssh/ssh_config\r\ndebug1: /etc/ssh/ssh_config line 58: Applying options for *\r\ndebug1: auto-mux: Trying existing master\r\ndebug2: fd 3 setting O_NONBLOCK\r\ndebug2: mux_client_hello_exchange: master version 4\r\ndebug3: mux_client_forwards: request forwardings: 0 local, 0 remote\r\ndebug3: mux_client_request_session: entering\r\ndebug3: mux_client_request_alive: entering\r\ndebug3: mux_client_request_alive: done pid = 13966\r\ndebug3: mux_client_request_session: session request sent\r\ndebug1: mux_client_request_session: master session id: 2\r\ndebug3: mux_client_read_packet: read header failed: Broken pipe\r\ndebug2: Received exit status from master 1\r\nShared connection to 213.1.1.79 closed.\r\n", "module_stdout": "sudo: a password is required\r\n", "msg": "MODULE FAILURE", "parsed": false}```

Any workaround known?

**Comment by Nirkus** [comment]
@mickgortenmulder @bbb31 
For me this issue turned out to be a server-side SSH connection multiplexing bug with Ubuntu 14.04.
Try to turn on debugging on the `sshd` side of the problematic conenction and look at the logs.

Ubuntu 14.04 seems to have broken server-side with their ConsoleKit integration, but I guess there are many other ways to break connection multiplexing. ;-)

https://bugs.launchpad.net/ubuntu/+source/openssh/+bug/1334916

**Comment by nacht** [comment]
**SOLVED** 

For anyone googling this - my particular thing was that I needed to add the user/group to the /etc/sudoers file with NOPASSWD set.

Have a similar issue with 

- Ubuntu 17.10 (GNU/Linux 4.13.0-16-generic x86_64)
- OpenSSH_7.5p1 Ubuntu-10, OpenSSL 1.0.2g

Have also tried with -f 1 and -c paramiko

fatal: [my.host]: FAILED! => {
    "changed": false,
    "failed": true,
    "module_stderr": "OpenSSH_7.4p1, LibreSSL 2.5.0\r\ndebug1: Reading configuration data /Users/me/.ssh/config\r\ndebug1: Reading configuration data /etc/ssh/ssh_config\r\ndebug1: auto-mux: Trying existing master\r\ndebug2: fd 3 setting O_NONBLOCK\r\ndebug2: mux_client_hello_exchange: master version 4\r\ndebug3: mux_client_forwards: request forwardings: 0 local, 0 remote\r\ndebug3: mux_client_request_session: entering\r\ndebug3: mux_client_request_alive: entering\r\ndebug3: mux_client_request_alive: done pid = 3524\r\ndebug3: mux_client_request_session: session request sent\r\ndebug1: mux_client_request_session: master session id: 2\r\ndebug3: mux_client_read_packet: read header failed: Broken pipe\r\ndebug2: Received exit status from master 1\r\nShared connection to my.host  closed.\r\n",
    "module_stdout": "sudo: a password is required\r\n",
    "msg": "MODULE FAILURE",
    "rc": 1
}

**Comment by ansibotdev** [comment]
@targetoyes You have not responded to information requests in this issue so we will assume it no longer affects you. If you are still interested in this, please create a new issue with the requested information.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)
